library/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt

```diff
@@ -80,52 +101,40 @@ class ChuckerInterceptor @JvmOverloads constructor(
                 transaction.isResponseBodyPlainText = false
             }
         }
+    }
 
-        collector.onRequestSent(transaction)
-
-        val startNs = System.nanoTime()
-        val response: Response
-        try {
-            response = chain.proceed(request)
-        } catch (e: IOException) {
-            transaction.error = e.toString()
-            collector.onResponseReceived(transaction)
-            throw e
-        }
-
-        val tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs)
-
+    /**
+     * Processes a [Response] and populates corresponding fields of a [HttpTransaction].
+     */
+    private fun processResponse(response: Response, transaction: HttpTransaction) {
         val responseBody = response.body()
+        val responseEncodingIsSupported = io.bodyHasSupportedEncoding(response.headers().get(CONTENT_ENCODING))
 
-        // includes headers added later in the chain
-        transaction.setRequestHeaders(filterHeaders(response.request().headers()))
         transaction.apply {
-            responseDate = System.currentTimeMillis()
-            this.tookMs = tookMs
+            // includes headers added later in the chain
+            setRequestHeaders(filterHeaders(response.request().headers()))
+            setResponseHeaders(filterHeaders(response.headers()))
+
+            isResponseBodyPlainText = responseEncodingIsSupported
+            requestDate = response.sentRequestAtMillis()
+            responseDate = response.receivedResponseAtMillis()
             protocol = response.protocol().toString()
             responseCode = response.code()
             responseMessage = response.message()
 
             responseContentType = responseBody?.contentType()?.toString()
             responseContentLength = responseBody?.contentLength() ?: 0L
-            setResponseHeaders(filterHeaders(response.headers()))
-        }
 
-        val responseEncodingIsSupported = io.bodyHasSupportedEncoding(response.headers().get(CONTENT_ENCODING))
-        transaction.isResponseBodyPlainText = responseEncodingIsSupported
+            tookMs = (response.receivedResponseAtMillis() - response.sentRequestAtMillis())
+        }
 
         if (response.hasBody() && responseEncodingIsSupported) {
             processResponseBody(response, responseBody, transaction)
         }
-
-        collector.onResponseReceived(transaction)
-
-        return response
     }
 
     /**
-     * Private method to process the HTTP Response body and populate the corresponding response fields
-     * of a the [HttpTransaction].
+     * Processes a [ResponseBody] and populates corresponding fields of a [HttpTransaction].
```

Comment:
Since you're refactoring this class, maybe you can extract the `.contains("image")` and move the string literal to the `companion object`?

Reply:
Sure, will do